src/coreComponents/physicsSolvers/fluidFlow/OBLSuperEngineKernels.hpp

```diff
@@ -0,0 +1,1507 @@
+      {
+        transMultDerI[v] = 0;
+        transMultDerJ[v] = 0;
+      }
+    }
+
+    real64 const pDiff = m_pres[erJ][esrJ][eiJ] + m_dPres[erJ][esrJ][eiJ] - m_pres[erI][esrI][eiI] - m_dPres[erI][esrI][eiI];
+
+
+    // [2] fill offdiagonal part + contribute to diagonal, only fluid part is considered in energy equation
+    for( integer p = 0; p < numPhases; p++ )
+    {     // loop over number of phases for convective operator
+
+      // calculate gravity term for phase p
+      real64 const averageDensity = (OBLValsI[GRAV_OP + p] + OBLValsJ[GRAV_OP + p]) / 2;
+
+      // p = 1 means oil phase, it's reference phase. pw=po-pcow, pg=po-(-pcog).
+      real64 const phasePDiff = (pDiff + averageDensity * gravCoef - OBLValsJ[PC_OP + p] + OBLValsI[PC_OP + p]) * pascalToBarMult;
+
+      // calculate partial derivatives for gravity and capillary terms
+      real64 gravPcDerI[numDofs];
+      real64 gravPcDerJ[numDofs];
+      for( integer v = 0; v < numDofs; v++ )
+      {
+        gravPcDerI[v] = -(OBLDersI[GRAV_OP + p][v]) * gravCoef / 2 - OBLDersI[PC_OP + p][v];
+        gravPcDerJ[v] = -(OBLDersJ[GRAV_OP + p][v]) * gravCoef / 2 + OBLDersJ[PC_OP + p][v];
+      }
+
+      real64 phaseGammaPDiff = transMult * trans * m_dt * phasePDiff;
+
+      if( phasePDiff < 0 )
+      {
+        // mass and energy outflow with effect of gravity and capillarity
+        for( integer c = 0; c < numDofs; c++ )
+        {
+          real64 compFlux = transMult * trans * m_dt * OBLValsI[FLUX_OP + p * numDofs + c];
+
+          stack.localFlux[c] -= phasePDiff * compFlux;     // flux operators only
+          // if( phasePDiff < 10 )
+          //   printf( "OUTFlux comp %d is %e subtracted %e trans %e m_dt %e\n", c, stack.localFlux[c], phasePDiff * compFlux, trans, m_dt
+          // );
+          for( integer v = 0; v < numDofs; v++ )
+          {
+            // if( eiI == 2 && v == 0 )
+            //   printf( "OUTFlux diag contrib0 %ld %ld is %e \n", eiI * numDofs + c, eiI * numDofs + v, stack.localFluxJacobian[c][v] );
+            stack.localFluxJacobian[c][v] -= (phaseGammaPDiff * OBLDersI[FLUX_OP + p * numDofs + c][v] +
+                                              trans * m_dt * phasePDiff * transMultDerI[v] * OBLValsI[FLUX_OP + p * numDofs + c]);
+            // if( eiI == 2 && v == 0 )
+            //   printf( "OUTFlux diag contrib1 %ld %ld is %e \n", eiI * numDofs + c, eiI * numDofs + v, stack.localFluxJacobian[c][v] );
+            stack.localFluxJacobian[c][v] += compFlux * gravPcDerI[v];
+            // if( eiI == 2 && v == 0 )
+            //   printf( "OUTFlux diag contrib2 %ld %ld is %e \n", eiI * numDofs + c, eiI * numDofs + v, stack.localFluxJacobian[c][v] );
+            stack.localFluxJacobian[c][numDofs + v] += compFlux * gravPcDerJ[v];
+
+            if( v == 0 )
+            {
+              stack.localFluxJacobian[c][numDofs + v] -= compFlux;
+              stack.localFluxJacobian[c][v] += compFlux;
+            }
```

Comment:
Maybe just add a quick comment here to say that this is the contribution to the dR_dP of the term corresponding to `compFlux` times the derivative of the pressure appearing in `phasePDiff` wrt pressure (and this last term is one), if I understand correctly.

Reply:
I don`t get what you mean here, sorry...